### PR TITLE
fix overfit_batch sampler replacement logic

### LINKED ISSUE #10457: `overfit_batches` does not work with `WeightedRandomSampler`
## 🐛 Bug

When the sampler `WeightedRandomSampler` is used in the dataloader, the option `overfit_batches` does not work properly as a new random batch will be loaded every time.

### To Reproduce

`WeightedRandomSampler` uses a multinomial distribution under the hood to sample data points given some weights. In the command line, if I set the seed, then I am able to sample the same points

```
>>> import pytorch_lightning as pl
>>> import torch
>>> pl.seed_everything(12); torch.multinomial(torch.tensor([0.1, 1.0, 1.0, 0.1, 2.0]), 3, False)
# tensor([2, 4, 1])
>>> pl.seed_everything(12); torch.multinomial(torch.tensor([0.1, 1.0, 1.0, 0.1, 2.0]), 3, False)
# tensor([2, 4, 1])
```
Here is the boring model https://colab.research.google.com/drive/1Q64RMumll0y0hQIFDmDeW8RaPs9_ryYD?usp=sharing
Notice that I print the batch every step, and this batch is always different:

```
0 tensor([[ 5.,  5.,  5.,  5.,  5.],
        [22., 22., 22., 22., 22.],
        [35., 35., 35., 35., 35.]])

0 tensor([[39., 39., 39., 39., 39.],
        [10., 10., 10., 10., 10.],
        [ 5.,  5.,  5.,  5.,  5.]])

0 tensor([[14., 14., 14., 14., 14.],
        [15., 15., 15., 15., 15.],
        [16., 16., 16., 16., 16.]])

0 tensor([[23., 23., 23., 23., 23.],
        [15., 15., 15., 15., 15.],
        [27., 27., 27., 27., 27.]])

0 tensor([[47., 47., 47., 47., 47.],
        [20., 20., 20., 20., 20.],
        [ 2.,  2.,  2.,  2.,  2.]])
```

### Expected behavior

The same batch should be used when over fitting.

### Environment

```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0+cu101
	- pytorch-lightning: 0.10.0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020
```

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #10457
The fix is that we check for `RandomSampler` to identify whether shuffling is happening or not.. instead we should check whether the sampler is `SequentialSampler` or not.
Also improved some similar tests a little bit

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -591,16 +590,17 @@ def _add_sampler_metadata_collate(dataloader: DataLoader) -> None:
...
     @staticmethod
     def _resolve_overfit_batches(dataloader: Collection[DataLoader]) -> Collection[DataLoader]:
-        has_random_sampler = False
+        all_have_sequential_sampler = True
 
-        def resolve_had_random_sampler(dataloader: DataLoader):
-            nonlocal has_random_sampler
-            if not has_random_sampler:
-                has_random_sampler = isinstance(dataloader.sampler, RandomSampler)
+        def resolve_has_no_sequential_sampler(dataloader: DataLoader):
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/data_loading.py:595]: yes, the check and replacement for `DistributedSampler` with happen right after this one. So basically replacement of samplers will happen twice in such a case. Do you think it's worth adding conditions to not replace with any sampler here if it's in distributed mode?

[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/data_loading.py:595]: I think it's okay to keep it as it is currently. The whole data loading part is due for a refactor. Let's keep this PR focused to the fix

File: tests/trainer/flags/test_overfit_batches.py [hunk 2]
```diff
@@ -31,25 +34,65 @@ def validation_epoch_end(self, outputs) -> None:
...
     )
 
     trainer.fit(model)
+    assert trainer.num_training_batches == overfit_batches
+    assert len(trainer.num_val_batches) == val_dl_count
+    assert all(nbatches == overfit_batches for nbatches in trainer.num_val_batches)
 
 
-@pytest.mark.parametrize("overfit", [1, 2, 0.1, 0.25, 1.0])
-def test_overfit_basic(tmpdir, overfit):
```
Comment (carmocca) [inline tests/trainer/flags/test_overfit_batches.py:48]: Funny how useless this test was! Thanks for the update

**Review by carmocca — APPROVED** [review]

File: tests/trainer/flags/test_overfit_batches.py (hunk 2 above)
Comment (rohitgr7) [reply tests/trainer/flags/test_overfit_batches.py:48]: yeah.. must be too old

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]
